src/Models/ArduinoDeviceBase.ts

```diff
@@ -232,41 +233,28 @@ export abstract class ArduinoDeviceBase implements Device {
     }
   }
 
-  async generateCrc(
-      context: vscode.ExtensionContext, channel: vscode.OutputChannel) {
+  async generateCrc(channel: vscode.OutputChannel): Promise<void> {
     if (!(vscode.workspace.workspaceFolders &&
           vscode.workspace.workspaceFolders.length > 0)) {
-      const message = 'No workspace opened.';
-      vscode.window.showWarningMessage(message);
-      utils.channelShowAndAppendLine(channel, message);
-      return false;
+      throw new Error('No workspace opened.');
     }
 
     const devicePath = ConfigHandler.get<string>(ConfigKey.devicePath);
     if (!devicePath) {
-      const message = 'No device path found in workspace configuration.';
-      vscode.window.showWarningMessage(message);
-      utils.channelShowAndAppendLine(channel, message);
-      return false;
+      throw new Error('No device path found in workspace configuration.');
     }
     const deviceBuildLocation = path.join(
         vscode.workspace.workspaceFolders[0].uri.fsPath, '..', devicePath,
         '.build');
 
     if (!deviceBuildLocation) {
-      const message = 'No device compile output folder found.';
-      vscode.window.showWarningMessage(message);
-      utils.channelShowAndAppendLine(channel, message);
-      return false;
+      throw new Error('No device compile output folder found.');
     }
 
     const binFiles = fs.listSync(deviceBuildLocation, ['bin']);
     if (!binFiles || !binFiles.length) {
-      const message =
-          'No bin file found. Please run the command of Device Compile first.';
-      vscode.window.showWarningMessage(message);
-      utils.channelShowAndAppendLine(channel, message);
-      return false;
+      throw new Error(

```

Comment:
define new class ResourceNotFoundError and use here

Reply:
Fixed.